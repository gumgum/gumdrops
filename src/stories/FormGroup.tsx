import React, { ReactChild, ReactChildren } from 'react';
import cx from 'classnames';
import { Colors } from 'types';

export interface FormGroupProps extends React.HTMLAttributes<Element> {
    color?: Colors;
    inline?: boolean;
    disabled?: boolean;
}

const baseClass = 'gds-form-group';

export const FormGroup: React.FC<FormGroupProps> = ({
    color = 'default',
    inline,
    disabled,
    className,
    children,
    ...otherProps
}) => {
    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${color}`]: color,
        [`${baseClass}--disabled`]: disabled,
        [`${baseClass}--inline`]: inline
    });

    return (
        <div className={rootClass} {...otherProps}>
            {children}
        </div>
    );
};
