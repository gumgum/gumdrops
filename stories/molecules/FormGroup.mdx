import { Canvas, Source } from '@storybook/blocks';
import { ArgsTable } from '@storybook/addon-docs';

import * as FormGroupStories from './FormGroup.stories';
import FormGroup from '../../components/molecules/FormGroup';
import FormGroupLabel from '../../components/atoms/FormGroupLabel';
import FormGroupTextHelp from '../../components/atoms/FormGroupTextHelp';

# FormGroup

The `<FormGroup>` component is used to wrap your form's inputs.

For a simple text input, it can consist of `<FormGroupLabel>`, `<TextInput>`, and `<FormGroupTextHelp>` components.

However, this depends on what type of input you are working on. Context is optional, and you probably don't need it unless you want to display a hint, warning, or error.

## TextInput example

```jsx
<FormGroup context="danger">
    <FormGroupLabel .../>
    <TextInput .../>
    <FormGroupTextHelp .../>
</FormGroup>
```

## Select example

```jsx
<FormGroup>
    <FormGroupLabel .../>
    <Select .../>
</FormGroup>
```

## Component Example

<Canvas of={FormGroupStories.Default} />

## Component Example Source

<Source of={FormGroupStories.Default} />

## FormGroup Props

<ArgsTable of={FormGroup} />

## FormGroupLabel Props

<ArgsTable of={FormGroupLabel} />

## FormGroupTextHelp Props

<ArgsTable of={FormGroupTextHelp} />
