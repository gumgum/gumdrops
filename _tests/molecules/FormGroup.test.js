import renderer from 'react-test-renderer';
import React from 'react';
import FormGroup from '../../components/molecules/FormGroup';

const defaultProps = {
    isInline: false,
    isDisabled: false,
    context: 'warning',
    className: 'custom-class'
};

test('Expect <FormGroup> to render properly', () => {
    const tree = renderer
        .create(
            <FormGroup {...defaultProps}>
                <input type="text" />
            </FormGroup>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test('Expect <FormGroup> to render disabled', () => {
    const props = {
        ...defaultProps,
        isDisabled: true
    };
    const tree = renderer
        .create(
            <FormGroup {...props}>
                <input type="text" />
            </FormGroup>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test('Expect <FormGroup> to render inline', () => {
    const props = {
        ...defaultProps,
        isInline: true
    };
    const tree = renderer
        .create(
            <FormGroup {...props}>
                <input type="text" />
            </FormGroup>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
