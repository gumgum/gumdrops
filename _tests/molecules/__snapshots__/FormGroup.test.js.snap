// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <FormGroup> to render disabled 1`] = `
<div
  className="gds-form-group custom-class gds-form-group--warning gds-form-group--disabled"
>
  <input
    type="text"
  />
</div>
`;

exports[`Expect <FormGroup> to render inline 1`] = `
<div
  className="gds-form-group custom-class gds-form-group--warning gds-form-group--inline"
>
  <input
    type="text"
  />
</div>
`;

exports[`Expect <FormGroup> to render properly 1`] = `
<div
  className="gds-form-group custom-class gds-form-group--warning"
>
  <input
    type="text"
  />
</div>
`;
